### ISSUE TITLE
Configuration improvement

### ISSUE THREAD

**Issue by silverhammermba** [awesomeWM/awesome#442]
`rc.lua` is one of the most awesome things about awesome. Just by changing one file you can turn awesome into an almost entirely different window manager. But despite this extreme flexibility, if you browse Linux desktop screenshots online it's generally pretty easy to pick out which ones use awesome. They all have that distinctive wibox with the taglist, tasklist, systray, clock, and a flat color theme. So it appears that even though awesome has the capability to completely change its appearance and behavior, there is a de facto "standard theme" which involves only certain superficial changes to the default `rc.lua`.

The problem with this is that `rc.lua` controls many different aspects of awesome and contains a mix of hardcoded drop-in values (e.g. tag names, menu items) and program logic (e.g. constructing window decorations, defining bind actions). Trying to merge features from someone else's `rc.lua` into your own can be incredibly difficult since they might have custom logic that you need to decipher in order to adapt it or determine it to be unnecessary. Take, for example, [awesome-copycats](https://github.com/copycat-killer/awesome-copycats) which is a bunch of different visual themes for awesome, where _each theme_ includes its own `rc.lua`! Why should changing the visual theme of awesome involve replacing the file that contains fundamental logic like showing errors on startup and binding the same shortcuts to each client? I think this really hurts the ease of customization of awesome by making it extremely difficult to share configs (or parts of configs) in a way that won't break someone else's setup. This problem comes up even for users who don't share configs, because if awesome's API changes in any way, your `rc.lua` will probably break unless you can incorporate the latest changes from the default one.

I think this situation can be improved in several ways:

1. Decide which features of the default `rc.lua` could be considered "standard" and add them to the Lua API so that we can have less logic in `rc.lua`. I think good candidates here are stuff like the error handling code, having common keybindings for tags/clients, having the same layouts per tag, common actions like task switching or viewing a single tag, etc. Basically the kind of stuff that pretty much everyone doesn't change.
2. To go one step further, add methods to the Lua API for easily modifying the config after the fact. So rather than making a copy of the default `rc.lua` and changing a few lines of code, you can just load the default as-is and then make a few more API calls to tweak the parts you don't like. For example, customizing key binds per-client, modifying the main wibox of one of your screens, defining different error handlers, etc. Those who want a more custom experience would just skip loading the default and build it all from scratch.
3. Break out more stuff into separate files a la `theme.lua`. Good candidates here would be stuff like key bindings, client rules, menu items; anything that is logically independent from other stuff. The API calls from 1 would help here, since there would be fewer custom callbacks to be coded. This would make it easier to drop in, say, a different visual appearance without worrying about changing other parts of your config.

This would require some pretty big changes, and we'd have to be careful to not break everyone's config. I can imagine it also being tricky to decide which features should be "standardized", and coming up with good APIs for allowing common customizations to be easy. But I honestly think that the mishmash of `rc.lua` is one of the biggest design problems with awesome.

**Comment by psychon** [comment]
Quick thought (this is basically just your idea 3, but taken to an extreme): Create a new lua module named `config` (similarly to `awful`) and the default rc.lua just does:

    local naughty = require("naughty")
    local config = require("config")
    naughty.init(whatever-the-default-config-does)
    config.error_popups()
    config.tags()
    config.key_bindings()
    for s = 1, screen.count() do
        config.wibox(s)
    end
    (etc)

So basically the default config would just be ~15 lines of code that loads various other stuff. Each of this (`config.tags`, `config.wibox`, ...) would be a separate file which contains everything that currently the default config contains. If you just want to overwrite the wibox, you create a file `~/.config/awesome/config/wibox.lua` and this file gets loaded instead of the one shipped with awesome.

If this split is done in a sane way *and* the various functions accept arguments (for example `config.key_bindings` could accept a list of extra key bindings that are *also* added), this would make it easier to overwrite specific parts of the default config. There are then the possibility of making use of the arguments that are provide or just overriding that part of the default "config" and replacing it completely with your own code.

So, what are the usual "suspects" for modifications? Based on this (plus some common sense) we should come up with a good split-up and a good API.

Oh and a quick thought: `config.wibox.extra_widgets` is just a (pretty much) empty file that can provide more widgets which then `config.wibox` incorporates somewhere. So if you just want to add a cpu-usage widget, you create it here and it gets included automatically.

**Comment by Bew78LesellB** [comment]
@silverhammermba I totally agree !
in fact, without noticing it first, I'm currently rewriting an API which aims to replace awful by a more modular approach , new concepts , etc .. (I can't say more ATM)

It will notably allow you to make a kind of default configuration, which could be customized by a file which will load extra module if needed, a bit like you described @psychon

**Comment by AlexisBRENON** [comment]
Hi totally agree too!

Reading awesome configuration from another people is very often a real nightmare, a big file with more than 500 loc, which mix many things from variable definition, widget creation and layout, etc.

I tried to separate all these concerns in my config, with a config folder that just define some (many) variables, a builder folder to create widgets selecting a standard, vicious or custom implementation, etc.

It's not perfect but maybe it can give some ideas !

**Comment by ZedCee** [comment]
I wouldn't mind managing a modular config setup. +1  Break a little less at a time.  
Would be nice if there was an error in one, it would load a default.  Maybe as a way to prevent breakage?
I'm no lua genius, rc's can get heavy, and complicated,; It possible that's not feasible for all of them.

**Comment by atakiel** [comment]
Modularized config sounds awesome.

Hope this also paves way to moving configs in the direction of making all configs modules/plugins. In a sense that all configs come as plugin, whether they are core or community made.

A thing worth considering, is to allow module loading happen in a separate config file, instead of only RC.lua. (Although, I'm not sure if it's possible already.) That would allow scripts to be used for loading additional modules. Perhaps something like package.json in javascript world, or I guess moonrocks in lua world.

It would be nice if some time in far future, adding custom modules to awesome wm would be as easy as adding plugins to github's atom editor. Mainly that it could be done without knowing any lua at all.

More or less all the config code in atom are done using plugins. There's just separation to official core plugins that ship with atom, and the plugins made by the rest of the community. But it's possible for user to enable / disable any and all of them.

A setup like that works very well in atom, and might be worth stealing/considering also for awesome wm.

Related to keybinds, atom has a nice hierarchial keybind system, where user only adds new keybinds or unsets lower ones in hierarchy (default or specified in some module). That way the custom keybinds file stays short, but is very powerful. System like that allows the default keybinds to change without breaking users custom keybinds (at least in most cases).

**Comment by mcgarc** [comment]
It seems that the main things to separate out are the keybindings and widgets, that might be what could be focused on to start with. It would then be a matter of moving all the logic somewhere else.

Might it also be worth incorporating ideas such as in #989 which suggests methods to simplify the way that keybindings are handled, there could then be a separate keybindings.lua file which can quickly and easily be modified directly.

Then as @atakiel suggests a more long-term goal could see custom modules overwriting these keybindings, perhaps as simply as being read after the suggested keybindings.lua 

Has anyone begun work on these features yet?

**Comment by AlexisBRENON** [comment]
Hi guys, 

I don't know if anybody have begun to work on it. Nevertheless, someone seems motivated on the [mailing list](http://www.mail-archive.com/awesome@naquadah.org/msg08243.html).

I would like also to take part in this development. I suggested that the mailing thread continues here to keep track of ideas and Co. 

Maybe the first steps are :
 * to define what must be set in different files
 * to define how each object (widget, keybinding, tag, etc.) can be represented/defined

**Comment by mcgarc** [comment]
At a first pass I would suggest that the things that should be in separate files are:

- A list of all keybindings
- Each individual widget

Perhaps the latter should include what currently exists in the rc.lua to create the taglist, etc. for each screen? This bit I'm a little unsure of.

I think this will ultimately allow us to create a clear and easily modified config, with the possibility of expanding to either automatically detect widgets that are placed in the relevant directory and define keybindings in some method similar to that mentioned above.

However I'm not sure of how we should go about representing these objects beyond what has already been discussed above. Perhaps someone else can offer an opinion on this?

**Comment by actionless** [comment]
i like the idea described by psychon:
https://github.com/awesomeWM/awesome/issues/442#issuecomment-139482282

it also allows to have few sets of default keybindings/panel layouts and making config extensions more standardized

are there any volunteers for implementing this already?
if not i can try

**Comment by mcgarc** [comment]
I have started attempting an implementation, I'll update here once I've made some headway.

**Comment by Ryuno-Ki** [comment]
@mcgarc Any updates? I discovered it via #1448.
Personally, I started with creating Lua files for the menu and include it in the rc.lua.

This helped me to better understand what's going on in rc.lua.

Maybe we should think about what are the essential bits, which must not be overwritten by plugins / modules. Things like „how to bind a key” etc (where the mapping can be overwritten).

**Comment by mcgarc** [comment]
Not really, I'm afraid. I got bogged down in some other stuff and never got back to this.

I think what would be good to start with is just continuing to modularise the other components of rc.lua and then work on each one individually. The keybindings module is probably a good example, you could start with what is essentially one file that is much the same as the keybindings section of rc.lua is now. The next step could potentially be to expand this module to two files, one with all the functions you wouldn't normally expect to change (like you say, "how to bind a key") and another with all the keybindings.

**Comment by Alechan** [comment]
I also gave it a try but stopped a few months ago. I decided to start again today.

My work in progress: https://github.com/Alechan/awesomerc_modular

My main goal is to have a modularized rc.lua that's 100% equivalent to the "monolithic" default.

One of my secondary goals is to have a clear distinction between user configuration and functionality. That is, that the functions, for example that generate the tags in the wibox, are in a separate file than the names, icons, etc; for example, the tag names "1.web", "2.terminal", etc.

Another secondary goal is to make it easily extensible and customizable. If I do things correctly, this should be a direct consequence of my main goal.

Feel free to contribute with issues, comments, pull requests, etc.
